onnxruntime/python/tools/transformers/onnx_model_tnlr.py

```diff
@@ -0,0 +1,179 @@
+        qkv_nodes = self.model.match_parent_path(start_node,
+                                                 ['Where', 'Add', 'MatMul', 'Reshape', 'Transpose', 'MatMul'],
+                                                 [1, 1, 1, 0, 0, 0])
+        if qkv_nodes is not None:
+            (_, _, matmul_below, reshape_qkv, transpose_qkv, matmul_qkv) = qkv_nodes
+        else:
+            return
+
+        other_inputs = []
+        for i, input in enumerate(start_node.input):
+            if input not in output_name_to_node:
+                continue
+
+            if input == qkv_nodes[0].output[0]:
+                continue
+            other_inputs.append(input)
+        if len(other_inputs) != 1:
+            return
+
+        root_input = other_inputs[0]
+
+        v_nodes = self.model.match_parent_path(matmul_qkv, ['Transpose', 'Reshape', 'Slice', 'Add', 'MatMul'],
+                                               [1, 0, 0, 0, 1])
+        if v_nodes is None:
+            return
+        (_, _, _, add, matmul) = v_nodes
+
+        upper_nodes = self.model.match_parent_path(matmul, ['Transpose'], [0])
+        transpose = upper_nodes[0]
+
+        qk_nodes = self.model.match_parent_path(matmul_qkv, ['Softmax', 'Add', 'MatMul'], [0, 0, 0])
+        if qk_nodes is None:
+            return
+        (_, add_qk, matmul_qk) = qk_nodes
+
+        q_nodes = self.model.match_parent_path(matmul_qk, ['Mul', 'Transpose', 'Reshape', 'Slice', 'Add', 'MatMul'],
+                                               [0, 0, 0, 0, 0, 1])
+        if q_nodes is None:
+            return
+        add = q_nodes[-2]
+        matmul = q_nodes[-1]
+
+        k_nodes = self.model.match_parent_path(matmul_qk, ['Transpose', 'Reshape', 'Slice', 'Add', 'MatMul'],
+                                               [1, 0, 0, 0, 1])
+        if k_nodes is None:
+            return
+        add = k_nodes[-2]
+        matmul = k_nodes[-1]
+
+        extra_add_qk_nodes = self.model.match_parent_path(add_qk, ['Reshape', 'Where'], [1, 0])
+        if extra_add_qk_nodes is None:
+            return
+
+        if matmul.input[0] == root_input:
+            mask_index = None
+            attention_last_node = reshape_qkv
+            # number of heads are same for all the paths, hence to create attention node, we pass the q_num_heads
+            # the input_hidden_size represents the input hidden size, this is used as needed but hidden sizes for Q, K are extracted appropriately
+            new_node = self.create_attention_node(mask_index, matmul, add, self.num_heads, self.hidden_size, root_input,
+                                                  attention_last_node.output[0], extra_add_qk_nodes[0].input[0])
+            if new_node is None:
+                return
+
+            self.nodes_to_add.append(new_node)
+            self.node_name_to_graph_name[new_node.name] = self.this_graph_name
+
+            # Add a transpose node after the attention in Offensive V4
```

Comment:
Do not put "Offensive V4" here. The term is not well known by general user. That will cause confusion.